numpy_financial/_financial.py

```diff
@@ -767,28 +741,23 @@ def irr(values):
     if values.ndim != 1:
         raise ValueError("Cashflows must be a rank-1 array")
 
-    # Strip leading and trailing zeros. Since we only care about
-    # positive roots we can neglect roots at zero.
-    non_zero = np.nonzero(np.ravel(values))[0]
-    values = values[int(non_zero[0]):int(non_zero[-1])+1]
+    solution_found = False
+    p = np.polynomial.Polynomial(values)
+    pp = p.deriv()
+    x = 1 / (1 + guess)
 
-    res = _roots(values[::-1])
+    for i in range(100):
+        x_new = x - (p(x) / pp(x))
+        if abs(x_new - x) < 1e-12:
+            solution_found = True
+            break
+        x = x_new
 
-    mask = (res.imag == 0) & (res.real > 0)
-    if not mask.any():
+    if solution_found:
+        return 1 / x - 1
+    else:
```

Comment:
This needs another check to ensure ``Decimal`` nan is returned when input datatype is ``Decimal``